librad/src/net/protocol/io/graft.rs

```diff
@@ -0,0 +1,193 @@
+        Refs(#[from] refs::stored::Error),
+
+        #[error("unable to obtain fetcher")]
+        Fetcher(#[from] fetcher::error::Retrying<git2::Error>),
+
+        #[error(transparent)]
+        Storage(#[from] storage::Error),
+
+        #[error(transparent)]
+        Pool(#[from] storage::PoolError),
+    }
+}
+
+pub mod config {
+    use super::*;
+
+    pub struct Rere {
+        pub replication: replication::Config,
+        pub fetch_slot_wait_timeout: Duration,
+    }
+}
+
+/// Initiate [`replication::replicate`] if the `remote_peer` appears to serve
+/// "interesting" refs.
+///
+/// Remote refs are deemed interesting if there is an intersection between the
+/// set of [`PeerId`]s acc. to the local [`Remotes`] for the given [`Urn`], and
+/// the advertised remote tracking branches of the remote peer (plus the remote
+/// peer itself).
+///
+/// This function is invoked when the `remote_peer` initiates a fetch (hence
+/// "rere" -- replicate replicate), as per [`super::recv::git`].
+///
+/// A fetch initiated by this function will **not** generate a
+/// [`crate::git::p2p::header::Header::nonce`], so as to not trigger reres
+/// recursively. Using this function thus requires to inspect the git header for
+/// the presence of a nonce (or else skip the rere), and to keep track of recent
+/// nonces in case of nonce re-use.
+#[tracing::instrument(level = "debug", skip(storage, config, addr_hints), err)]
+pub async fn rere<S, Addrs>(
+    storage: S,
+    config: config::Rere,
+    urn: Urn,
+    remote_peer: PeerId,
+    addr_hints: Addrs,
+) -> Result<Option<replication::ReplicateResult>, error::Rere>
+where
+    S: storage::Pooled + Send + Sync + 'static,
+    Addrs: IntoIterator<Item = SocketAddr>,
+{
+    fetcher::retrying(
+        storage,
+        fetcher::PeerToPeer::new(urn.clone(), remote_peer, addr_hints).nonced(false),
+        config.fetch_slot_wait_timeout,
+        move |storage, fetcher| {
+            let remote_heads = fetcher.remote_heads();
+            let refs = Refs::load(&storage, &urn, None)
+                .map_err(error::Rere::from)?
+                .ok_or_else(|| error::Rere::MissingSignedRefs(urn.clone()))?;
+
+            tracing::debug!("remotes: {:?}", refs.remotes);
+
+            if is_interesting(&remote_peer, &remote_heads, &refs.remotes) {
+                tracing::debug!("interesting");
+                Ok(Some(
+                    replication::replicate(&storage, fetcher, config.replication, None)
+                        .map_err(error::Rere::from)?,
+                ))
+            } else {
+                tracing::debug!("uninteresting");
+                Ok(None)
+            }
+        },
+    )
+    .await?
+}
+
+fn is_interesting<P>(remote_peer: &P, remote_heads: &RemoteHeads, remotes: &Remotes<P>) -> bool
+where
+    P: Ord + FromStr,
+{
+    let tracked_local = remotes.flatten().collect::<BTreeSet<_>>();
+    //
+    // Ok, computer science NERDS. Can we make this actually efficient?
```

Comment:
Since we're just looking for _not being disjoint_, there is the opportunity to short-circuit with something like `any`. That is, we iterate through the `refs`, parse, and check if it's in `tracked_local`. That should be linear in `refs` and has the best-case of exiting as early as possible.

However, is this not an opportunity for us to _know_ which `PeerId`s two sides share and that means we can focus replication on those only?

Reply:
> to short-circuit with something like any

Ah yeah... not sure why I thought there could be duplicates in the input.

> However, is this not an opportunity for us to know which PeerIds two sides share and that means we can focus replication on those only?

But we can't make use of this information -- `replicate` figures out which remotes it wants, so knowing that it will find at least one is enough.

Reply:
> But we can't make use of this information -- replicate figures out which remotes it wants, so knowing that it will find at least one is enough.

We can't at the moment, but I'm thinking we could allow `replicate` to be intialised with a set of peers that's interesting for the fetch. Maybe that's not an improvement though?

Reply:
Yes, we would save re-reading the tracked peers when computing the `Peek` fetchspecs. That's not nothing, but also not super-critical. I thought we might want to apply more algebra to the replication procedure first :nerd_face:

Reply:
Sounds good :smile: I better `solve for x` so